### DISCUSSION TITLE
Polling rate? can we do better than 12s?

### DISCUSSION THREAD

**Discussion by tmenguy** [wlcrs/huawei_solar#718] [General]
Hi,

I'm on the latest 1.4 beta, SDongle via wired ethernet for connection. I use an automation to update the power meter and inverter active power sensor:

- alias: "Huawei Solar power data polling"
  trigger:
    - platform: time_pattern
      hours: "*"
      minutes: "*"
      seconds: "/4"
  action:
    - service: homeassistant.update_entity
      target:
        entity_id: 
            - sensor.power_meter_active_power
            - sensor.inverter_active_power

and it seems I can't go really beyond 12s (looking at the update graphs and the concrete state update of the sensors) Is there someone that can go below? should I change the polling frequencies directly in the const.py ? I would like to get closer to 5s to implement some control algorithms.

Thanks!
BR
Thomas

**Comment by Roving-Ronin** [comment]
@tmenguy 
...hold my beer ;-)
Seriously though, bit of a how long is a piece of string / can I nitro boost my car situation. How fast is also going to depend upon how many devices you have in your plant. i.e. If you only a single inverter + power meter, you might be able to go fast.... if your like my setup with 2 x Inverters + Meter + 38 Optimisers it will be a bit slower, as it needs to poll 41 devices with 484 entities between them.

Someone (you) mentioned getting 15 seconds previously, I'd be happy with that.... and remember on the flip side the HA Recorder is going to be saying all that data away and the Db is going to get big and slow.

**Comment by tmenguy** [comment]
Hi,

(first did I say that I love your avatar?) 

I would like to implement a kind of EVCC like stuff directly as a custom component to handle my car smart charging ... AND my soon to come battery + some of my loads like pools, etc ( I have a Wallbox with a meter and ecosmart but it probably won't work well with the Huawei optimisation algorithms)

I have 40 optimizers , meter and one inverter (and soon a battery), so not so far from your setup.

By putting directly 5s in the const.py of the integration, I'm now down to 6 or 7 seconds ... but I still don't understand why it is faster this way than using the automations with trigger,  @wlcrs if you have an idea?

I may go down a bit to in the code to see if I can squeeze a bit more calls, but I don't know what is the theoretical Modbus limits I may hit (and the real update period of the meter itself).

**Reply by wlcrs** [reply]
No idea. Mu guess is that the HA automation trigger process is not optimized for such high frequency updates.

Starting from v1.4.0 the integration will only poll the sensors that are effectively enabled in HA. You could look into disabling all but the most essential sensors of this integration and check what the performance you get that way. Afterwards you can try re-enabling sensors a few at a time to check what the performance hit is. Feel free to report back your findings :-)

**Reply by Roving-Ronin** [reply]
Hi Thijs,

So is there any way to disable all Optimisers, or not have them discovered/installed, to cut down on the amount of polling?

**Comment by tmenguy** [comment]
yes, thx a lot, I'll fiddle a bit and do a fork to see if we can have something of a higher frequency "for free" (like the Shelly interaction that is more around the second for their refresh)
Good point for the device/entity we can disable, will look at it too, thx

**Comment by tmenguy** [comment]
interesting findings, here:
-disabled all optimizers : no impact , and it is normal the new 1.40 is really well decoupled on this, the optimizers are pull every 5mn so no real impact
-I missed how the HA polling was implemented in the DataUpdateCoordinator but it is pretty simple : once you have finished your update you add the polling interval to the now, so you can't go faster than real polling interval = [your fetching and update time] + [setup polling interval] (I probably learn you nothing @wlcrs here)

When I set the update to 1s as polling interval I have a result of around a data point every 3s, probably meaning it takes for me around 2s to gather the data from the SDongle ... which seems a lot to me to be honest (it is a bit less to be honest here from my tests more between 1 and 2s, still not insignificant)

All in all I have a solution here :) 3 to 4s of update is plenty fast for what I need

@wlcrs : would it be worthwhile to you if I open a PR to add an option config flow to your component to select a polling interval in the UI configuration, or you think it is too dangerous for general use? (and if so I'll just hardcode stuff for my personal use)

**Reply by PulsarFX** [reply]
is this with the const.py modification? 
I tried to update 4 entities on the 1.4.1 @2s intervals, but the values update only every 10s.

**Reply by tmenguy** [reply]
Yes by modifying directly the const.py, and putting 1s in it for the polling of :

INVERTER_UPDATE_INTERVAL = timedelta(seconds=1)
POWER_METER_UPDATE_INTERVAL = timedelta(seconds=1)
ENERGY_STORAGE_UPDATE_INTERVAL = timedelta(seconds=1)
UPDATE_TIMEOUT = timedelta(seconds=29)

not on the inverters nor anything else, and I disabled a lot of entities I don't need too.

I'm between 3 and 5s update on my system, and the HA is running in a VM on a Mac mini with a good M2 CPU

**Reply by PulsarFX** [reply]
Hm. I changed the values, restarted HA and tested the fast automation with 2s and 5s.  The GUI shows updates every 10s, though. On 2s the trigger header is always orange, while 5s has most of the time a blue header.
As I am reading only 4 sensors, traces shows execution times of only 0.0x seconds.

**Reply by tmenguy** [reply]
Hey, which « GUI » are you referring to?

**Reply by heinemannj** [reply]
> @wlcrs : would it be worthwhile to you if I open a PR to add an option config flow to your component to select a polling interval in the UI configuration, or you think it is too dangerous for general use? (and if so I'll just hardcode stuff for my personal use)

To be honest and crystal clear:

- The mayority of huawei_solar users have already activated **battery control via installer authorization** - t**he most important benefit and feature of this beautiful integration**!!!
- For controlling and changing inverter/batteries system parameters you (all user itself) are responsible for any change of system parameters
- Everybody using this integration must have an approbiate level of knowledge about interverter/batteries operations

=> Reading an additional little peace of documentation about polling parameters is peanuts compared to above ...

@wlcrs: PLEASE accept this merge request - It makes for the majority of your followers their life a little bit more easier.

**Reply by PulsarFX** [reply]
> Hey, which « GUI » are you referring to?

ops, overlooked this one. I mean the sensors detail page, like:
![grafik](https://github.com/user-attachments/assets/f5dbca98-0a86-4caf-bf73-0e4e5f4633f1)

it always resets at 10s, whilst the automation with call to `service: homeassistant.update_entity`  is set to trigger each 5s.

**Reply by PulsarFX** [reply]
Found the issue: I had disabled polling by the integration and used my own automations instead. 
With them disabled and default polling by the integration enabled, I get about real time updates to 5s updates. 
I am using:
```
INVERTER_UPDATE_INTERVAL = timedelta(seconds=1)
POWER_METER_UPDATE_INTERVAL = timedelta(seconds=1)
ENERGY_STORAGE_UPDATE_INTERVAL = timedelta(seconds=1)
UPDATE_TIMEOUT = timedelta(seconds=29)
```

(in the meantime I managed to setup modbuxproxy and query against that)

i.e. the dtsu666 sensor is updating like crazy. 8-)

**Comment by LotharWoman** [comment]
I found this : https://solaranzeige.de/phpBB3/viewtopic.php?t=4400
Can this workaround be used here to make the queries faster and more reliable?

**Comment by heinemannj** [comment]
Polling via automation as proposed by @wlcrs is also not working fine for me.
**Modifying directly the const.py is much more faster and more reliable**.

To archive small update frequcencies without any HA error logs, I had to switch my network connection directly to the inverter WLAN (not to the dongle) => **this helps a lot**!!!

BUT after the switch to from `v1.3.x` to `v1.4.x` I recognize time issues: https://github.com/wlcrs/huawei_solar/issues/767

![image](https://github.com/user-attachments/assets/fa0ee22c-2152-4c50-843a-3730fb876f16)

To make this story short:

`house_power: "{{ inv_active_power - pm_active_power }}"`

**Negative House Load** (for my setup NOT possible) - potentially due to Poll/Timing mismatch between `Inverter Active power` and `Power meter Active power`:

**NOT recognized in versions 1.3.x**.

**My proposal**:

- dedicated update interval for all power related sensors (inverter, power-meter, batteries): interval 1-10s
- dedicated update interval for all energy related sensors (inverter, power-meter, batteries): interval 10-30s
- for remaining sensors (inverter, power-meter, batteries): interval 30-60s

**Any ideas or proposals for the power sensors**???

For the energy and all other sensor's it's not an issue.

**Comment by spamadamm** [comment]
I've managed stable refresh rate of 3 seconds for EMMA 02 + inverter. I don't have the battery at the moment. I suggest to increase Modbus RTU baudrate from 9600 to 19200, maybe that improves the pooling rate over Modbus TCP a bit. It has to be done on all your devices in order to work.

**Reply by LotharWoman** [reply]
Changing the baud rate is not a solution. It can reset due to configuration changes or updates. In the worst case scenario, various components no longer communicate with each other.

**Comment by tmenguy** [comment]
I submitted a PR for this (#828 ) for those who wants to test something, you can test with my repo (see the PR), it is based on the latest (as of today) code, so far so good for me on my install.

**Reply by Roving-Ronin** [reply]
@tmenguy Is that updated off the latest (v1.4.1) code, or v1.5.1 Alpha 1 ?

Edit: Assuming v.1.5.1 Alpha 1

Appears to work:

![image](https://github.com/user-attachments/assets/25ec203e-9b81-4cfe-b015-0981997953d2)

**Reply by heinemannj** [reply]
Unfortunately after installation of your repository the integrations goes into error and stops.
The first step of configuration workflow seems to be OK.

I will make additional debugging when I have a couple of hours free time.

**Reply by tmenguy** [reply]
@Roving-Ronin yes it is latest code base of the repo, while update it while we test and @wlcrs is ok (or not) to merge it

@heinemannj : where you on 1.4.1? Because the 1.5.x has some auto discovery code for the slaveIDs for me the "AUTO" for the slaveID was not working I had to put back my slaveID (1 in my case) and then it works. If you were in 1.5.x are you in serial or network? (I haven't tested serial....)


And by the way thanks you both for testing!

**Reply by heinemannj** [reply]
Actual I'm on the origional 1.5.0a1.
After understanding that this new version reacts different to the older versions **I was** also **successful with your patched version**! 

![image](https://github.com/user-attachments/assets/e2f3c055-44cb-4367-afe0-1e3eb6bbf7cc)

![image](https://github.com/user-attachments/assets/6a004123-7966-419d-b264-a056f40cdf86)

![image](https://github.com/user-attachments/assets/fb6269c1-e67a-4972-b9a9-7727d37cd395)

![image](https://github.com/user-attachments/assets/d3b4c046-1094-4bee-85c1-322f814c3dde)

**Polling frequencies for all of the three devices as expected.**

Power meter: 3sec
Inverter: 3-6sec
Batteries: 3-8sec

Increasing to 4 or 5 secs seems to be a good  compromise.

**Thanks for your work!**

Actually I'm working on an prototype for an HEMS: https://github.com/heinemannj/huawei_solar_hems
Polling rates at 30sec are far away from the need of an Home Assistent (HA) Home Energy Management System (HEMS).

By the way:
I'm using the 3sec polling rates since more then 9 months without any problems.

**Reply by tmenguy** [reply]
As an HEMS you can look at mine : tmenguy/quiet-solar it is working for some months for me, EV charging, pool, some loads, etc

**Reply by Peethiplatsch** [reply]
> Actual I'm on the origional 1.5.0a1. After understanding that this new version reacts different to the older versions **I was** also **successful with your patched version**!
> 
> ![image](https://private-user-images.githubusercontent.com/3251642/376939826-e2f3c055-44cb-4367-afe0-1e3eb6bbf7cc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg4NTc2MTksIm5iZiI6MTc0ODg1NzMxOSwicGF0aCI6Ii8zMjUxNjQyLzM3NjkzOTgyNi1lMmYzYzA1NS00NGNiLTQzNjctYWZlMC0xZTNlYjZiYmY3Y2MucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MDJUMDk0MTU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDA2NWU3NmFjZjg5NmRiY2JlYzAwMDAyMTU2MjBmNjczM2Y5ZjRmY2UzYTQ3ZWJkY2JlYzE3YTI5ZTc5NzA0ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.2hjVQKTCx-l5p0VHuKZJXrvITMiKi8uI5xJAnxc0cio)
> 
> ![image](https://private-user-images.githubusercontent.com/3251642/376950774-6a004123-7966-419d-b264-a056f40cdf86.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg4NTc2MTksIm5iZiI6MTc0ODg1NzMxOSwicGF0aCI6Ii8zMjUxNjQyLzM3Njk1MDc3NC02YTAwNDEyMy03OTY2LTQxOWQtYjI2NC1hMDU2ZjQwY2RmODYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MDJUMDk0MTU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZWNmNTY0ZGVmZTFiNjViMzc4YTJiM2ZhZmEzMjdkMzNmN2Q0ZGI0ZTUxMjk4N2FiZjk5YmRlMjFjYzM4M2NlNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.VcEWeISQ89hBD5Mj1bBa0hYso04DenPYRVnHrffMlVM)
> 
> ![image](https://private-user-images.githubusercontent.com/3251642/376950985-fb6269c1-e67a-4972-b9a9-7727d37cd395.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg4NTc2MTksIm5iZiI6MTc0ODg1NzMxOSwicGF0aCI6Ii8zMjUxNjQyLzM3Njk1MDk4NS1mYjYyNjljMS1lNjdhLTQ5NzItYjlhOS03NzI3ZDM3Y2QzOTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MDJUMDk0MTU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDU3OGZhZjk2Y2U0ZjQ5MjhiM2RmZmU5ZWFiNWRlMTU4YTQ1OTA5ZWFhOTA3ODcwNDA5MTZjNzAxNjNkMTdhZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.dqB1WUkKZFFfoXkpAOBu2cMV3VQjz3-9zqmYr5zbE44)
> 
> ![image](https://private-user-images.githubusercontent.com/3251642/376943399-d3b4c046-1094-4bee-85c1-322f814c3dde.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg4NTc2MTksIm5iZiI6MTc0ODg1NzMxOSwicGF0aCI6Ii8zMjUxNjQyLzM3Njk0MzM5OS1kM2I0YzA0Ni0xMDk0LTRiZWUtODVjMS0zMjJmODE0YzNkZGUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDYwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA2MDJUMDk0MTU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmEyOTA0NDU0MmRiMGY2YWJmZWJkZTJmYWJhOTk1Zjg3NjRlM2M5YWVkNGUxY2UxZTY4NWFkYjYxYjhmNTY4OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ap2ctnv6Nn_2_rJXJuMoAoxe2Qz3zixJNPiLeO9Fabc)
> 
> **Polling frequencies for all of the three devices as expected.**
> 
> Power meter: 3sec Inverter: 3-6sec Batteries: 3-8sec
> 
> Increasing to 4 or 5 secs seems to be a good compromise.
> 
> **Thanks for your work!**
> 
> Actually I'm working on an prototype for an HEMS: https://github.com/heinemannj/huawei_solar_hems Polling rates at 30sec are far away from the need of an Home Assistent (HA) Home Energy Management System (HEMS).
> 
> By the way: I'm using the 3sec polling rates since more then 9 months without any problems.

Hey its a bit late to ask. I have installed the 1.5.4 version and don't have the options to change the polling. What have you done to get this on the configuration screen? Would love to have faster polling as well :)

**Reply by tmenguy** [reply]
I continue to maintain this : https://github.com/tmenguy/huawei_solar to have those settings, closely following latest version of the official one

**Comment by LotharWoman** [comment]
I think with the new firmware 121 for the sDongle, no adjustment is necessary anymore.

![grafik](https://github.com/user-attachments/assets/53bfe7fa-6aee-47ae-a32b-3b15d2f88c3d)

**Reply by tmenguy** [reply]
Hum the polling rate is on the HA code part, and not dependent on  the SDongle firmware to me, so not sure how to get a faster refresh without changing the HA part.

**Reply by spamadamm** [reply]
Unfortunately, it is not possible to use the SDongle with the EMMA 02 energy management system, so we still search for a solution.

**Comment by heinemannj** [comment]
**Unfortunately the PR has been rejected by @wlcrs**

As already stated above:

- Polling via automation as proposed by @wlcrs is not working fine for me.
- Modifying directly the const.py is much more faster and more reliable.
- Using above patch will simplyfy a lot and is better supportable.

My recommendation:

- The lowest polling time should be limited to 3 sec (From my perspective lower then 3sec makes no sense)

**Reply by Roving-Ronin** [reply]
:-(   Hopefully Thijs reconsiders this...?

This PR is something that would be beneficial, leaves the default (30s) settings for those only needing the standard option...but easily and cleanly allows those in need of fasting polling to easily/cleanly have that option.

Having this as an option also ensures those needing it have a way of doing it, whilst avoiding 'fiddling' with .py files and things 'under the covers', thereby reducing any chance they edit something incorrectly and cause an issue that traces back to a bad .py edit. At the sametime 

I can understand Thijs not wanting to officially support those setting extremely low poll intervals and as such, if this was included in the integration, it should have a disclaimer that anyone modifying the poll frequency from the default (or at least under say 20s) will not be supported / will need to change back to default 30s if they have a problem before they lodge and issue and submit debug logs etc.

Edit: Re the comment 'people trying to get the poll frequency too low'... @tmenguy maybe the PR could you modified to set a minimum frequency that could be set? 15 - 30s ?

**Reply by heinemannj** [reply]
Minimum frequency at 15s is to slow for my need.
As mentioned above: ~5 seconds should work in most cases

**Reply by Roving-Ronin** [reply]
Yes but there's a point that Thijs will has to consider will allow 'most' people to go down to and not potentially run into issues. Decision has to be made based upon all people, not just yourself.

On the flipside though, if this was approved with a (for example) 15s-30s limit, this would still at least allow you to edit 1 file and change 1 line of code to change it down to your 5s frequency, thereby still allowing you to benefit by not having to try Automations or editing multiple .py files (and do so after every update).

Still a win - win, even with 15-30s (and update your own limits in a single .py to 5-30s thereafter).

**Reply by heinemannj** [reply]
You are confusing me ...

There is already a stable workarround by modifying directly the const.py because it is much more faster and more reliable then using HA automations to decrease the Huawei Solar polling frequency.

Disadvantage of this workarround: You will need to apply this hack again and again - after each integration update ...

**What file needs to be modified after applying above patch** to go below predefined (silly) limits?

I've found nothing at this time.
Please clarify

**Reply by Roving-Ronin** [reply]
I was referring to IF this PR was included people would not have to di the workaround. It would only be (very few) people who would need to modify a version that includes this PR, after official updates, to solely (unofficially) change the update frequency minimum limitation (solely for their own HA instance).

**Reply by heinemannj** [reply]
For the moment we have the file: `const.py`
**Which file needs to be modified in future**???

If HEMS is common in future => to secure our earth => Demand of just in time actions will be increased!

**Reply by tmenguy** [reply]
No hard feelings it is Thijs code, I closed the PR will see if I prefer to maintain my repo, or revert back to edit may const.py :)

**Reply by heinemannj** [reply]
From my perspective maintaining your repro really make sense - the frequency of new Huawei Solar versions is not to high at this time.

**Comment by henfri** [comment]
Hello,

it seems that V126 significantly improves the Modbus Stability:
https://www.loxforum.com/forum/german/software-konfiguration-programm-und-visualisierung/443912-verbesserung-der-kommunikation-ms-wr-nach-huawei-pv-update

Greetings,
Hendrik

**Reply by t785634** [reply]
Has anyone tested the new SDongle version V200R022C10SPC200 yet?

**Comment by kvogel456** [comment]
hi there, 

anybody can tell me something about polling rates (or response time) with **const.py modifications** with EMMA connected to LAN ?

my setup now is:
connected to inverter with wlan bridge, with const.py tweak im getting data about every 3-4 seconds

EMMA ready to install, but im afraid of getting a "downgrade" in response time?

**Reply by tmenguy** [reply]
no downgrade for me with Emma (with my custom repo)

**Reply by kvogel456** [reply]
how often do you get "fresh" data?

**Reply by tmenguy** [reply]
3s

**Reply by kvogel456** [reply]
one more question:
do u have battery in your setup? if so, can it be controlled via EMMA from this integration?

**Reply by tmenguy** [reply]
hey yes I do have a battery, and yes I can.  Strangely to control it, in HA, it is not part of the EMMA entities, but still on the battery for a part, and another part not inverter ... but anyway it works
